File: haskell/Day03.hs
module Day03 where

import Text.Regex.TDFA
import Data.List (isPrefixOf)

findALl :: String -> String -> [String]
findALl inp reg = getAllTextMatches (inp =~ reg) :: [String]

mulProc :: String -> Int
mulProc i = foldl (\acc x -> acc * (read x)) 1 (findALl i "[0-9]+")

processor :: [String] -> Bool -> Int
processor [] _ = 0
processor (x:xs) activated
    | isPrefixOf "don't" x = processor xs False 
    | isPrefixOf "do(" x = processor xs True
    | otherwise = (if activated then (mulProc x) else 0)  + processor xs activated

solve :: String -> String
solve inp = mconcat ["Part 1: ", p1s, "\n", "Part 2: ", p2s]
    where
        p1s = show $ sum $ map mulProc $ findALl inp  "mul\\([0-9]+,[0-9]+\\)"
        p2s = show $ processor (findALl inp  "mul\\([0-9]+,[0-9]+\\)|do\\(\\)|don't\\(\\)") True

File: python/day03.py
import re


def parser(test):
    return open(f"./data/day03/{"test" if test else "inp"}.txt", "r").read().strip()


def part_1(test: bool = False):
    regex = re.compile(r"mul\(\d+.\d+\)")
    tot = 0
    for case in regex.findall(parser(test)):
        procd = re.compile(r"\d+").findall(case)
        tot += int(procd[0]) * int(procd[1])
    return tot


def part_2(test: bool = False):
    regexb = re.compile(r"(mul\(\d+.\d+\))|(do\(\))|(don't\(\))")
    tot = 0
    enabled = True
    for case in regexb.findall(parser(test)):
        if case[1] != "":
            enabled = True
        elif case[2] != "":
            enabled = False
        else:
            procd = re.compile(r"\d+").findall(case[0])
            if enabled:
                tot += int(procd[0]) * int(procd[1])
    return tot

if __name__ == "__main__":
    print(part_1())
    print(part_2())
